File: src/main/java/org/code4everything/hutool/converter/ListStringConverter.java
package org.code4everything.hutool.converter;

import cn.hutool.core.io.FileUtil;
import cn.hutool.core.util.StrUtil;
import cn.hutool.http.HttpUtil;
import org.code4everything.hutool.Converter;
import org.code4everything.hutool.Hutool;
import org.code4everything.hutool.MethodArg;
import org.code4everything.hutool.Utils;

import java.io.File;
import java.util.Collection;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.StringJoiner;

/**
 * @author pantao
 * @since 2020/11/10
 */
public class ListStringConverter implements Converter<List<String>> {

    public boolean directLineSep = false;

    public ListStringConverter useLineSep() {
        directLineSep = true;
        return this;
    }

    @Override
    public List<String> string2Object(String string) {
        if (Utils.isStringEmpty(string)) {
            return Collections.emptyList();
        }

        if (string.startsWith("file:")) {
            File file = new FileConverter().string2Object(string.substring(5));
            Hutool.debugOutput("get file path: " + file.getAbsolutePath());
            if (FileUtil.exist(file)) {
                string = FileUtil.readUtf8String(file);
            }
        } else if (string.startsWith("http:") || string.startsWith("https:")) {
            string = HttpUtil.get(string);
        }

        if (directLineSep) {
            return StrUtil.splitTrim(string, "\n");
        }

        String separator = MethodArg.getSeparator();
        if (!separator.contains("\n")) {
            string = StrUtil.strip(string, "[", "]");
        }
        return StrUtil.splitTrim(string, separator);
    }

    @Override
    public String object2String(Object object) throws Exception {
        if (!(object instanceof Collection)) {
            return "";
        }

        Collection<?> list = (Collection<?>) object;
        if (Utils.isCollectionEmpty(list)) {
            return "[]";
        }

        Iterator<?> iterator = list.iterator();
        if (list.size() == 1) {
            return Hutool.convertResult(iterator.next(), null);
        }

        StringJoiner joiner;
        String separator = MethodArg.getSeparator();
        if (directLineSep || separator.contains("\n")) {
            joiner = new StringJoiner(separator);
        } else {
            joiner = new StringJoiner(",", "[", "]");
        }

        while (iterator.hasNext()) {
            joiner.add(Hutool.convertResult(iterator.next(), null));
        }
        return joiner.toString();
    }
}

File: src/main/kotlin/org/code4everything/hutool/converter/ListStringConverter.kt
package org.code4everything.hutool.converter

import cn.hutool.core.io.FileUtil
import cn.hutool.core.util.StrUtil
import cn.hutool.http.HttpUtil
import java.util.StringJoiner
import org.code4everything.hutool.Converter
import org.code4everything.hutool.Hutool
import org.code4everything.hutool.MethodArg

class ListStringConverter : Converter<List<String?>> {

    private var directLineSep = false

    fun useLineSep(): ListStringConverter = apply { directLineSep = true }

    override fun string2Object(string: String): List<String> {
        if (string.isEmpty()) {
            return emptyList()
        }

        var str = string
        if (str.startsWith("file:")) {
            val file = FileConverter().string2Object(str.substring(5))
            Hutool.debugOutput("get file path: %s", file.absolutePath)
            if (FileUtil.exist(file)) {
                str = FileUtil.readUtf8String(file)
            }
        } else if (str.startsWith("http:") || str.startsWith("https:")) {
            str = HttpUtil.get(str)
        }

        if (directLineSep) {
            return StrUtil.splitTrim(str, "\n")
        }

        val separator = MethodArg.separator
        if (!separator.contains("\n")) {
            str = StrUtil.strip(str, "[", "]")
        }
        return StrUtil.splitTrim(str, separator)
    }

    override fun object2String(any: Any?): String {
        if (any !is Collection<*>) {
            return ""
        }
        if (any.isEmpty()) {
            return "[]"
        }

        val iterator = any.iterator()
        if (any.size == 1) {
            return Hutool.convertResult(iterator.next(), null)
        }

        val joiner: StringJoiner
        val separator = if (directLineSep) FileUtil.getLineSeparator() else MethodArg.separator
        joiner = if (separator.contains("\n")) StringJoiner(separator) else StringJoiner(",", "[", "]")

        while (iterator.hasNext()) {
            joiner.add(Hutool.convertResult(iterator.next(), null))
        }
        return joiner.toString()
    }
}
